### PR TITLE
Update to cellular automaton stochastic convective organization scheme

### PR DESCRIPTION
**Pull request by lisa-bengtsson**
# PR Checklist

- [x] Ths PR is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR. Please consult the ufs-weather-model [wiki](https://github.com/ufs-community/ufs-weather-model/wiki/Making-code-changes-in-the-UFS-weather-model-and-its-subcomponents) if you are unsure how to do this.

- [x] This PR has been tested using a branch which is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR

- [x] An Issue describing the work contained in this PR has been created either in the subcomponent(s) or in the ufs-weather-model. The Issue should be created in the repository that is most relevant to the changes in contained in the PR. The Issue and the dependent sub-component PR 
are specified below.

- [x] If new or updated input data is required by this PR, it is clearly stated in the text of the PR.

## Instructions: All subsequent sections of text should be filled in as appropriate.

The information provided below allows the code managers to understand the changes relevant to this PR, whether those changes are in the ufs-weather-model repository or in a subcomponent repository. Ufs-weather-model code managers will use the information provided to add any applicable labels, assign reviewers and place it in the Commit Queue. Once the PR is in the Commit Queue, it is the PR owner's responsiblity to keep the PR up-to-date with the develop branch of ufs-weather-model. 

## Description

Updates include:

1) Introduction of CA domain definition.
2) Read/Write and restart capability of CA
3) Change namelist control parameter so that CA time and space scales are adjusted for various resolutions given the same CA namelist input. 
4) Significant cleaning of old ideas. 
5) Improved coupling to cumulus convection to be committed in coordination with this issue: NCAR/ccpp-physics#657

Test results with GFSv16 as a baseline (atmosphere only) can be found here for winter and summer: https://www.emc.ncep.noaa.gov/gmb/jhan/vsdbw/v16upcas2/ and https://www.emc.ncep.noaa.gov/gmb/jhan/vsdbw/v16upcaw2

New baselines are needed for the fv3_ca regression tests. 

### Issue(s) addressed

Link the issues to be closed with this PR, whether in this repository, or in another repository.
(Remember, issues must always be created before starting work on a PR branch!) 
- fixes #<issue_number>
- fixes noaa-emc/fv3atm/issues/<issue_number>

## Testing

How were these changes tested? What compilers / HPCs was it tested with? Are the changes covered by regression tests? (If not, why? Do new tests need to be added?) Have regression tests and unit tests (utests) been run? On which platforms and with which compilers? (Note that unit tests can only be run on tier-1 platforms)

- [x] hera.intel
- [x] hera.gnu
- [x] orion.intel
- [x] cheyenne.intel 
- [x] cheyenne.gnu
- [x] gaea.intel 
- [x] jet.intel
- [x] wcoss_cray
- [x] wcoss_dell_p3
- [x] CI

## Dependencies

If testing this branch requires non-default branches in other repositories, list them. Those branches should have matching names (ideally).

Do PRs in upstream repositories need to be merged first?
If so add the "waiting for other repos" label and list the upstream PRs

https://github.com/noaa-psd/stochastic_physics/pull/41
https://github.com/NOAA-EMC/fv3atm/pull/305

### CODE REVIEW

**Review by climbfuji — COMMENTED** [review]
Please add links/references to the associated PRs in the PR description. Thanks!

File: tests/parm/control_ca.nml.IN [hunk 1]
```diff
@@ -181,15 +181,24 @@ deflate_level=1
...
-  nlives         = @[NLIVES]
-  nseed          = @[NSEED]
-  nthresh        = @[NTHRESH]
-  ca_trigger     = .true.
-  ca_smooth      = @[CA_SMOOTH]
-  nspinup        = @[NSPINUP]
-  iseed_ca       = @[ISEED_CA]
-  nsmooth        = @[NSMOOTH]
-  ca_amplitude   = @[CA_AMPLITUDE]
+  nca            = 1
```
Comment (junwang-noaa) [inline tests/parm/control_ca.nml.IN:184]: I want to confirm the updated setting will only be used when do_ca is .true., correct? The RT test control_lndp is also using the control_ca.nml.IN as namelist template, just want to make sure there is no impact on that test.

Comment (lisa-bengtsson) [reply tests/parm/control_ca.nml.IN:184]: That is correct, thanks for checking. Results only change if do_ca = True. New baselines are needed for the tests coupled and atmosphere only tests:

Test 004 cpld_ca 
Test 026 control_ca

File: tests/rt.sh [hunk 2]
```diff
@@ -240,12 +240,13 @@ elif [[ $MACHINE_ID = hera.* ]]; then
   QUEUE=debug
   COMPILE_QUEUE=batch
 
-  #ACCNR=fv3-cpu
+  ACCNR=ome
```
Comment (climbfuji) [inline tests/rt.sh]: The changes to `rt.sh` will need to be reverted.

Comment (lisa-bengtsson) [reply tests/rt.sh]: Yes, I will create new baselines, update rt.sh with a new date and revert these updates.

**Comment by MinsukJi-NOAA** [comment]
@lisa-bengtsson's RT on Hera resulted in two tests failing as it should:
control_ca 026
cpld_ca 004

I am going to generate baselines on all platforms now. @DeniseWorthen is generating baselines on Cheyenne

**Comment by BrianCurtis-NOAA** [comment]
Machine: cheyenne
Compiler: gnu
Job: RT
Repo location: /glade/scratch/dtcufsrt/autort/tests/auto/pr/649388260/20210611103009/ufs-weather-model
Please manually delete: /glade/scratch/dtcufsrt/FV3_RT/rt_21627
Test fv3_esg_HAFS_v0_hwrf_thompson_debug 026 failed in run_test failed 
Test fv3_HAFS_v0_hwrf_thompson_debug 025 failed in run_test failed 
Test cpld_debug 028 failed in run_test failed 
Test fv3_rrfs_v1alpha 011 failed failed 
Test fv3_rrfs_v1alpha 011 failed in run_test failed 
Test fv3_HAFS_v0_hwrf_thompson 013 failed in run_test failed 
Test fv3_esg_HAFS_v0_hwrf_thompson 014 failed in run_test failed 
Test fv3_rrfs_v1beta 012 failed in run_test failed 
Test fv3_gsd 010 failed in run_test failed 
Test fv3_rrfs_v1beta_debug 018 failed in run_test failed 
Test control_ugwpv1 008 failed in run_test failed 
Test control_ras_debug 024 failed in run_test failed 
Test control 001 failed in run_test failed 
Test control_ras 009 failed in run_test failed 
Test control_flake 004 failed in run_test failed 
Test fv3_rrfs_v1alpha_debug 017 failed in run_test failed 
Test fv3_gsd_debug 019 failed in run_test failed 
Test control_debug 015 failed in run_test failed 
Test regional_control_debug 016 failed in run_test failed 
Test control_thompson_debug 020 failed in run_test failed 
Test control_rrtmgp 005 failed in run_test failed 
Test control_thompson_no_aero_debug 021 failed in run_test failed 
Test control_rrtmgp_debug 022 failed in run_test failed 
Test control_thompson 006 failed in run_test failed 
Test control_ugwpv1_debug 023 failed in run_test failed 
Test control_thompson_no_aero 007 failed in run_test failed 
Test control_stochy 003 failed in run_test failed 
Test cpld_control 027 failed in run_test failed 
Please make changes and add the following label back:
cheyenne-gnu-RT

**Comment by DeniseWorthen** [comment]
A test which should not have changed (control_flake) shows this error for Cheyenne GNU:

```  
shoc_parm=   7000.0000000000000        1.0000000000000000        4.2857143000000004       0.69999999999999996       -999.00000000000000       ncnvw=           6  ncnvc=        -999
0:GFS_phys_time_vary_init: compute sncovr from weasd and soil vegetation parameters
0:
0:Program received signal SIGSEGV: Segmentation fault - invalid memory reference.
0:
0:Backtrace for this error:
```

**Comment by MinsukJi-NOAA** [comment]
All CI tests (regional_control and cpld_control) failed as well.
```
#3  0x56544bb3f224 in __stochastic_physics_wrapper_mod_MOD_stochastic_physics_wrapper
	at /home/builder/ufs-weather-model/FV3/stochastic_physics/stochastic_physics_wrapper.F90:282
#3  0x55a4bcdd42e4 in __stochastic_physics_wrapper_mod_MOD_stochastic_physics_wrapper
	at /home/builder/ufs-weather-model/FV3/stochastic_physics/stochastic_physics_wrapper.F90:284
#3  0x561cffc89284 in __stochastic_physics_wrapper_mod_MOD_stochastic_physics_wrapper
	at /home/builder/ufs-weather-model/FV3/stochastic_physics/stochastic_physics_wrapper.F90:283
```

**Comment by DeniseWorthen** [comment]
Yes, the same line appears for the control_debug :

```
14:#1  0x217f49a in __stochastic_physics_wrapper_mod_MOD_stochastic_physics_wrapper
14:     at /glade/scratch/dtcufsrt/autort/tests/auto/pr/649388260/20210611103009/ufs-weather-model/FV3/stochastic_physics/stochastic_physics_wrapper.F90:282
```

**Comment by MinsukJi-NOAA** [comment]
Since GNU tests are failing, I will wait to see if it's also the case with hera.gnu.

**Comment by lisa-bengtsson** [comment]
Perhaps I need to move the do_ca check (if do_ca) above the check:

if (GFS_Control%kdt==0) then

       do nb=1,Atm_block%nblks

          GFS_Data(nb)%Intdiag%ca_deep(:)  = 0.

          GFS_Data(nb)%Intdiag%ca_turb(:)  = 0.

          GFS_Data(nb)%Intdiag%ca_shal(:)  = 0.

          GFS_Data(nb)%Coupling%ca_deep(:) = 0.

          GFS_Data(nb)%Coupling%ca_turb(:) = 0.

          GFS_Data(nb)%Coupling%ca_shal(:) = 0.

          GFS_Data(nb)%Coupling%ca1(:) = 0.

          GFS_Data(nb)%Coupling%ca2(:) = 0.

          GFS_Data(nb)%Coupling%ca3(:) = 0.

          GFS_Data(nb)%Intdiag%ca1(:)  = 0.

          GFS_Data(nb)%Intdiag%ca2(:)  = 0.

          GFS_Data(nb)%Intdiag%ca3(:)  = 0.

       enddo

    else


?
---------------------------------
---------------------------------
Dr. Lisa Bengtsson
Phone : 1 (303) 497 5971
Research Scientist - Meteorologist
University of Colorado/NOAA Earth
System Research Lab
Email: ***@***.***
325 Broadway
Office: Skaggs Research Cntr 1D-702
Boulder, CO, USA 80303-3328

**Comment by lisa-bengtsson** [comment]
I can update and push the stochastic_physics_wrapper.F90, can you give me the command to test the gnu compiler on hera? Thanks!

**Comment by MinsukJi-NOAA** [comment]
> I can update and push the stochastic_physics_wrapper.F90, can you give me the command to test the gnu compiler on hera? Thanks!

Try this:
```
cd tests
export RT_COMPILER=gnu
./rt.sh -l rt_gnu.conf -e >out 2>&1 &
```

**Comment by MinsukJi-NOAA** [comment]
Just to confirm that Hera gnu jobs are failing as well.

**Comment by BrianCurtis-NOAA** [comment]
Machine: hera
Compiler: gnu
Job: RT
Repo location: /scratch1/NCEPDEV/nems/emc.nemspara/autort/pr/649388260/20210611171515/ufs-weather-model
Test fv3_rrfs_v1beta 012 failed failed 
Test fv3_rrfs_v1beta 012 failed in run_test failed 
Test fv3_gsd 010 failed failed 
Test fv3_gsd 010 failed in run_test failed 
Test fv3_rrfs_v1alpha 011 failed failed 
Test fv3_rrfs_v1alpha 011 failed in run_test failed 
Test cpld_debug 028 failed failed 
Test cpld_debug 028 failed in run_test failed 
Test fv3_HAFS_v0_hwrf_thompson_debug 025 failed failed 
Test fv3_HAFS_v0_hwrf_thompson_debug 025 failed in run_test failed 
Test fv3_HAFS_v0_hwrf_thompson 013 failed failed 
Test fv3_HAFS_v0_hwrf_thompson 013 failed in run_test failed 
Test fv3_esg_HAFS_v0_hwrf_thompson_debug 026 failed failed 
Test fv3_esg_HAFS_v0_hwrf_thompson_debug 026 failed in run_test failed 
Test fv3_gsd_debug 019 failed failed 
Test fv3_gsd_debug 019 failed in run_test failed 
Test control_debug 015 failed failed 
Test control_debug 015 failed in run_test failed 
Test fv3_esg_HAFS_v0_hwrf_thompson 014 failed failed 
Test fv3_esg_HAFS_v0_hwrf_thompson 014 failed in run_test failed 
Test regional_control_debug 016 failed failed 
Test regional_control_debug 016 failed in run_test failed 
Test control_ugwpv1_debug 023 failed failed 
Test control_ugwpv1_debug 023 failed in run_test failed 
Test fv3_rrfs_v1alpha_debug 017 failed failed 
Test fv3_rrfs_v1alpha_debug 017 failed in run_test failed 
Test control_ugwpv1 008 failed failed 
Test control_ras_debug 024 failed failed 
Test control_ugwpv1 008 failed in run_test failed 
Test control_ras_debug 024 failed in run_test failed 
Test control 001 failed failed 
Test control 001 failed in run_test failed 
Test control_thompson_no_aero 007 failed failed 
Test control_thompson_no_aero 007 failed in run_test failed 
Test control_rrtmgp 005 failed failed 
Test control_rrtmgp 005 failed in run_test failed 
Test fv3_rrfs_v1beta_debug 018 failed failed 
Test fv3_rrfs_v1beta_debug 018 failed in run_test failed 
Test control_ras 009 failed failed 
Test control_ras 009 failed in run_test failed 
Test control_thompson 006 failed failed 
Test control_thompson 006 failed in run_test failed 
Test control_thompson_no_aero_debug 021 failed failed 
Test control_thompson_no_aero_debug 021 failed in run_test failed 
Test cpld_control 027 failed failed 
Test cpld_control 027 failed in run_test failed 
Test control_rrtmgp_debug 022 failed failed 
Test control_rrtmgp_debug 022 failed in run_test failed 
Test control_flake 004 failed failed 
Test control_flake 004 failed in run_test failed 
Test control_stochy 003 failed failed 
Test control_stochy 003 failed in run_test failed 
Test control_thompson_debug 020 failed failed 
Test control_thompson_debug 020 failed in run_test failed 
Please make changes and add the following label back:
hera-gnu-RT

**Comment by lisa-bengtsson** [comment]
I updated the stochastic physics wrapper and running hera gnu test now, sitting in queue...

**Comment by DeniseWorthen** [comment]
@lisa-bengtsson Do we know if this will change answers for the two new baselines I created for cpld_ca and control_ca tests for intel?

**Comment by lisa-bengtsson** [comment]
I updated the stochastic_physics_wrapper.F90 and ran the hera gnu test which passed on Hera.
@DeniseWorthen to address your question - it shouldn't, but I've been surprised before... should I rerun the intel test on Hera?

**Comment by DeniseWorthen** [comment]
@lisa-bengtsson Minsuk suggested I go ahead and regenerate them on Cheyenne since it was only two. So the RT is running there now. Thanks.

**Comment by lisa-bengtsson** [comment]
> @lisa-bengtsson Minsuk suggested I go ahead and regenerate them on Cheyenne since it was only two. So the RT is running there now. Thanks.

Ok, thanks!

**Comment by BrianCurtis-NOAA** [comment]
Machine: hera
Compiler: intel
Job: RT
Repo location: /scratch1/NCEPDEV/nems/emc.nemspara/autort/pr/649388260/20210611203040/ufs-weather-model
Please manually delete: /scratch1/NCEPDEV/stmp2/emc.nemspara/FV3_RT/rt_180294
Test cpld_bmark_wave_v16 012 failed failed 
Test cpld_bmark_wave_v16 012 failed in run_test failed 
Please make changes and add the following label back:
hera-intel-RT

**Comment by BrianCurtis-NOAA** [comment]
Machine: gaea
Compiler: intel
Job: RT
Repo location: /lustre/f2/pdata/ncep/emc.nemspara/autort/pr/649388260/20210612114511/ufs-weather-model
Please manually delete: /lustre/f2/scratch/emc.nemspara/FV3_RT/rt_26174
Test control_ca 024 failed in check_result failed 
Test control_ca 024 failed in run_test failed 
Test cpld_ca 004 failed in check_result failed 
Test cpld_ca 004 failed in run_test failed 
Please make changes and add the following label back:
gaea-intel-RT

**Comment by lisa-bengtsson** [comment]
> Machine: gaea
> Compiler: intel
> Job: RT
> Repo location: /lustre/f2/pdata/ncep/emc.nemspara/autort/pr/649388260/20210612114511/ufs-weather-model
> Please manually delete: /lustre/f2/scratch/emc.nemspara/FV3_RT/rt_26174
> Test control_ca 024 failed in check_result failed
> Test control_ca 024 failed in run_test failed
> Test cpld_ca 004 failed in check_result failed
> Test cpld_ca 004 failed in run_test failed
> Please make changes and add the following label back:
> gaea-intel-RT

Hm... not sure what do make of this? What would be different on gaea? Is it possible to remake the baselines and run the regression test again?

**Comment by MinsukJi-NOAA** [comment]
@lisa-bengtsson I deleted your comment by mistake. Reproduced below with a response.

> Hm... not sure what do make of this? What would be different on gaea? Is it possible to remake the baselines and run the regression test again? 

This should be taken care of once the new baselines for *_ca are moved to the right place.

**Comment by lisa-bengtsson** [comment]
> @lisa-bengtsson I deleted your comment by mistake. Reproduced below with a response.
> 
> > Hm... not sure what do make of this? What would be different on gaea? Is it possible to remake the baselines and run the regression test again?
> 
> This should be taken care of once the new baselines for *_ca are moved to the right place.

Ok, thanks for clarifying!

**Comment by MinsukJi-NOAA** [comment]
This PR is ready for merge once the FV3 and stochastic_physics are merged in their repositories, and then the .gitmodules pointers are updated.

**Review by pjpegion — APPROVED** [review]

**Review by junwang-noaa — APPROVED** [review]

**Comment by lisa-bengtsson** [comment]
@junwang-noaa @MinsukJi-NOAA .gitmodules has been reverted.

**Comment by MinsukJi-NOAA** [comment]
> @junwang-noaa @MinsukJi-NOAA .gitmodules has been reverted.

@lisa-bengtsson, your FV3 hash should be 0a47bd1, not 453f38e. Also, stochastic_physics should be c142e67, not 2ba2421

**Comment by lisa-bengtsson** [comment]
> > @junwang-noaa @MinsukJi-NOAA .gitmodules has been reverted.
> 
> @lisa-bengtsson, your FV3 hash should be 0a47bd1, not 453f38e. Also, stochastic_physics should be c142e67, not 2ba2421

@MinsukJi-NOAA Ok, I thought I only needed to update .gitmodules, should I fetch upstream to the merged FV3 and stochastic_physics directories?

**Comment by MinsukJi-NOAA** [comment]
> @MinsukJi-NOAA Ok, I thought I only needed to update .gitmodules, should I fetch upstream to the merged FV3 and stochastic_physics directories?

Yes

**Comment by lisa-bengtsson** [comment]
> > @MinsukJi-NOAA Ok, I thought I only needed to update .gitmodules, should I fetch upstream to the merged FV3 and stochastic_physics directories?
> 
> Yes

Apologies that I don't get how this work. I tried to do "git fetch upstream", then: 
[Lisa.Bengtsson@hfe03 ufs-weather-model]$ git pull upstream develop
From https://github.com/ufs-community/ufs-weather-model
 * branch            develop    -> FETCH_HEAD
Already up to date.
[Lisa.Bengtsson@hfe03 ufs-weather-model]$ git status
On branch feature/CA_develop2
Your branch is up to date with 'origin/feature/CA_develop2'.

nothing to commit, working tree clean

Do I have to specify a number explicitly? If you could provide me the commands to point to these merged repositories I would greatly appreciate it.

**Comment by lisa-bengtsson** [comment]
@MinsukJi-NOAA @junwang-noaa am I pointing to the right place now?

**Comment by MinsukJi-NOAA** [comment]
> @MinsukJi-NOAA @junwang-noaa am I pointing to the right place now?

@lisa-bengtsson , FV3 is still pointing to a different hash. Try this:
```
$ cd FV3
$ git checkout develop
$ cd ..
$ git diff FV3
-Subproject commit 7a57a2bc70cf6777a4eedb415d91e0bd347cc55f
+Subproject commit 0a47bd13ab49a785e9731eb8d59a46b429f26720
```
git add, commit then push

**Comment by MinsukJi-NOAA** [comment]
FV3 is now correctly pointing to 0a47bd1.

**Comment by lisa-bengtsson** [comment]
Ok, thanks, I did this, does it look correct now?

**Pull request merged**